### PR TITLE
DEV: Preload all models schema cache before running system tests

### PR DESCRIPTION
**Pull request by tgxworld**
### Why this change?

When running system tests with all official plugins installed, we have
encountered instances where the system tests will hang. When dumping the
backtraces of the threads, we can see that the main thread running the
tests is stuck in a deadlock with the puma thread while serving a
request.

The deadlock happens when the main thread acquires the `ActiveSupport::Concurrency::LoadInterlockAwareMonitor`
lock first in `ActiveRecord::ConnectionAdapters::AbstractAdapter` before acquring another `Monitor` lock in
`ActiveRecord::ModelSchema`. In the Puma thread, it acquires the
`Monitor` lock in `ActiveRecord::ModelSchema` first before acquring the
`ActiveSupport::Concurrency::LoadInterlockAwareMonitor` lock.

### What does this change do?

To workaround this problem, we will preload all model schema cache
before running system tests such that the `Monitor` lock in `ActiveRecord::ModelSchema`
will not be acquired.

### CODE REVIEW

File: spec/rails_helper.rb [hunk 1]
```diff
@@ -499,8 +484,37 @@ class SpecTimeoutError < StandardError
   # Match the request hostname to the value in `database.yml`
   config.before(:each, type: %i[request multisite system]) { host! "test.localhost" }
 
-  last_driven_by = nil
```
Comment (tgxworld) [inline spec/rails_helper.rb:502]: This is an unused variable.

File: spec/rails_helper.rb [hunk 2]
```diff
@@ -239,21 +239,6 @@ def fab!(name, **opts, &blk)
...
-    # the same location and this can interfere with the running tests in another process.
-    #
-    # The long term fix here is to get `selenium-manager` to download the `chromedriver` binary to a unique path for each
-    # process but the `--cache-path` option for `selenium-manager` is currently not supported in `selenium-webdriver`.
-    if !File.directory?("~/.cache/selenium")
-      File.open("#{Rails.root}/tmp/chrome_driver_flock", "w") do |file|
-        file.flock(File::LOCK_EX)
-        `#{Selenium::WebDriver::SeleniumManager.send(:binary)} --browser chrome`
-      end
-    end
```
Comment (tgxworld) [inline spec/rails_helper.rb:255]: Moving this so we don't run it unnecessarily for non system tests.

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -499,8 +484,37 @@ class SpecTimeoutError < StandardError
...
+      # On Rails 7, we have seen instances of deadlocks between the lock in [ActiveRecord::ConnectionAdapaters::AbstractAdapter](https://github.com/rails/rails/blob/9d1673853f13cd6f756315ac333b20d512db4d58/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L86)
+      # and the lock in [ActiveRecord::ModelSchema](https://github.com/rails/rails/blob/9d1673853f13cd6f756315ac333b20d512db4d58/activerecord/lib/active_record/model_schema.rb#L550).
+      # To work around this problem, we are going to preload all the model schemas before running any system tests so that
+      # the lock in ActiveRecord::ModelSchema is not acquired at runtime. This is a temporary workaround while we report
+      # the issue to the Rails.
+      ActiveRecord::Base.connection.data_sources.map do |table|
+        ActiveRecord::Base.connection.schema_cache.add(table)
+      end
+
+      init_system_test = true
```
Comment (martin-brennan) [inline spec/rails_helper.rb:515]: Nitpick, but `init_system_test` reads backwards to me in its current form; it starts of false when I would expect it to be true, we want it for a "past tense" thing, so I think `system_tests_initialized` would be better :)

**Review by martin-brennan — APPROVED** [review]
Just added a minor comment

File: spec/rails_helper.rb (hunk 3 above)
Comment (tgxworld) [reply spec/rails_helper.rb:515]: Good point :+1: I have changed it to `system_tests_initialized`.

**Comment by tgxworld** [comment]
Thank you for the review @martin-brennan :+1:

**Pull request merged**